pkg/controllers/policyrc/controller.go

```diff
@@ -20,175 +19,143 @@ package policyrc
 import (
 	"context"
 	"fmt"
-	"strings"
 	"time"
 
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
-	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
-	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	pkgruntime "k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/klog/v2"
 
 	fedcorev1a1 "github.com/kubewharf/kubeadmiral/pkg/apis/core/v1alpha1"
 	"github.com/kubewharf/kubeadmiral/pkg/client/generic"
+	fedinformers "github.com/kubewharf/kubeadmiral/pkg/client/informers/externalversions"
+	fedcorev1a1informers "github.com/kubewharf/kubeadmiral/pkg/client/informers/externalversions/core/v1alpha1"
 	"github.com/kubewharf/kubeadmiral/pkg/controllers/common"
 	"github.com/kubewharf/kubeadmiral/pkg/controllers/override"
-	"github.com/kubewharf/kubeadmiral/pkg/controllers/scheduler"
 	"github.com/kubewharf/kubeadmiral/pkg/controllers/util"
-	"github.com/kubewharf/kubeadmiral/pkg/controllers/util/delayingdeliver"
-	"github.com/kubewharf/kubeadmiral/pkg/controllers/util/worker"
 	"github.com/kubewharf/kubeadmiral/pkg/stats"
+	"github.com/kubewharf/kubeadmiral/pkg/util/fedobjectadapters"
+	"github.com/kubewharf/kubeadmiral/pkg/util/logging"
+	"github.com/kubewharf/kubeadmiral/pkg/util/worker"
 )
 
 const (
 	ControllerName = "policyrc-controller"
 )
 
-var PolicyrcControllerName = common.DefaultPrefix + "policyrc-controller"
-
-type informerPair struct {
-	store      cache.Store
-	controller cache.Controller
-}
-
 type Controller struct {
-	// name of controller: <federatedKind>-policyrc-controller
-	name string
-
-	// Informer store and controller for the federated type, PropagationPolicy,
-	// ClusterPropagationPolicy, OverridePolicy and ClusterOverridePolicy respectively.
-	federated, pp, cpp, op, cop informerPair
-
-	client generic.Client
+	fedObjectInformer                fedcorev1a1informers.FederatedObjectInformer
+	clusterFedObjectInformer         fedcorev1a1informers.ClusterFederatedObjectInformer
+	overridePolicyInformer           fedcorev1a1informers.OverridePolicyInformer
+	clusterOverridePolicyInformer    fedcorev1a1informers.ClusterOverridePolicyInformer
+	propagationPolicyInformer        fedcorev1a1informers.PropagationPolicyInformer
+	clusterPropagationPolicyInformer fedcorev1a1informers.ClusterPropagationPolicyInformer
 
 	ppCounter, opCounter *Counter
 
 	// updates the local counter upon fed object updates
-	countWorker worker.ReconcileWorker
+	countWorker worker.ReconcileWorker[common.QualifiedName]
 	// pushes values from local counter to apiserver
-	persistPpWorker, persistOpWorker worker.ReconcileWorker
-
-	typeConfig *fedcorev1a1.FederatedTypeConfig
+	persistPpWorker, persistOpWorker worker.ReconcileWorker[common.QualifiedName]
 
+	client  generic.Client
```

Comment:
Could we use a typed client here?

Reply:
The generic client is currently used to reduce the boilerplate code needed to handle the 4 different policy types. We can revisit this if and when we choose to deprecate the generic client.